model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java

```diff
@@ -278,7 +280,19 @@ public MapStorageProvider create(KeycloakSession session) {
     }
 
     protected EntityManager getEntityManager() {
-        return emf.createEntityManager();
+        EntityManager em = emf.createEntityManager();
+
+        if ("postgresql".equals(databaseShortName) || "cockroachdb".equals(databaseShortName)) {
```

Comment:
Instead of asking Liquibase, there is existing code in other places (for example for the DB specific named queries), where we ask the DB connection about the product name. So there would be no need to use the information from Liquibase, and to expose it in the API/store it in attributes. 

See JpaMapUtils:

```
            String dbProductName = em.unwrap(Session.class).doReturningWork(connection -> connection.getMetaData().getDatabaseProductName());
            String dbKind = getDatabaseType(dbProductName);
``` 

I suggest to use only one way to determine the type of the database.

Reply:
Using Liquibase was suggested by @hmlnarik to not communicate with DB as it is not necessary since Liquibase already knows what DB is used.

Reply:
Last time I checked getting the name of the database from the JDBC driver was a local operation not involving the database (at least for the PostgreSQL and Oracle driver). So it might even return "PostgreSQL" for the CockroachDB.

I just noticed it, please proceed as you see fit.

Reply:
I will leave this decision to @hmlnarik as this was his idea.

Reply:
Liquibase distinguishes CRDB from PgSQL. Putting a breakpoint here https://github.com/keycloak/keycloak/blob/8d05895adb06d8b44f9447e057e26e265cb1efda/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/liquibase/connection/DefaultLiquibaseConnectionProvider.java#L74
and calling `database.getShortName()` results in `"cockroachdb"` in case of CRDB and `"postgresql"` in case of PgSQL.
Thus we can rely on Liquibase, and should not rely on JDBC metadata for the very arguments as posted by @ahus1.

Reply:
Should we create an issue to replace the code @ahus1 posted in JpaMapUtils with using databaseShortName? Or should I do it as part of this PR?

Reply:
Please file an issue for that

Reply:
Created: [#16617](https://github.com/keycloak/keycloak/issues/16617)